packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2865,8 +2861,11 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
             assert(!!maybeSnapshot, 0x138 /* "Failed to get snapshot from storage" */);
             stats.getSnapshotDuration = trace.trace().duration;
 
+            const attributes = await this.storage.getDocumentAttributesFromTree?.(maybeSnapshot);
```

Comment:
I'm a bit confused here RE why we need new method on storage interface, and what is the meaning of sequence number s in there. Drivers, for most part, should not have knowledge about attributes and treat is as any other blob.
I think today attributes concept is a loader concept (though storage knows about it for sure), and ideally it should stay that way, unless we want to expose non-virtualized sequence numbers to runtime.

Can you please share a bit more on what you are trying to accomplish here?

Reply:
So the issue here is that code has access to protocol tree that has non-virtualized data. I.e. leakage.
I think we have to figure out how to prevent that leakage and/or patch data in a way where no runtime changes are required.
ODSP driver used to do it (there is still code artifacts from those days, I believe). I think loader layer (virtual storage layer) has to do it, by creating "virtual blob" and patching tree with updated .attributes blob.
That has some negative implications (we introduce fake blob IDs, and theoretically runtime can reuse them). That said, it's easier to implement and its overall pretty solid, so we can start here and consider switching to other design (with more invasive changes) later.

Maybe better way - remove protocol tree from tree when passing it to runtime. And provide explicit APIs to get this data. What if we modify ISnapshotTree itself to include this data directly? I.e. introduce ISnapshotTreeRoot that extends ISnapshotTree and has decoded data from protocol, but omits protocol tree itself? That way loader can patch it while providing to runtime.

Reply:
Currently, the runtime-utils is doing this document attributes extraction (see runtime-utils/src/utils.ts).

I agree, the surface can probably be simplified in terms of the runtime not reaching into the tree for the values and instead being provided them explicitly. I think doing it in a more explicit way is safer than modifying the tree because it does not require sync on both sides (loader and runtime) in terms of the structure of the object.